control-plane-operator/hostedclusterconfigoperator/controllers/drainer/drainer.go

```diff
@@ -0,0 +1,162 @@
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	kubeclient "k8s.io/client-go/kubernetes"
+	"k8s.io/kubectl/pkg/drain"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+const (
+	// DesiredDrainerAnnotationKey is set by the MCD to indicate drain/uncordon requests
+	DesiredDrainerAnnotationKey = "machineconfiguration.openshift.io/desiredDrain"
+	// LastAppliedDrainerAnnotationKey is set by the controller to indicate the last request applied
+	LastAppliedDrainerAnnotationKey = "machineconfiguration.openshift.io/lastAppliedDrain"
+	// DrainerStateUncordon is used for drainer annotation as a value to indicate needing an uncordon
+	DrainerStateUncordon = "uncordon"
+	// DrainerStateDrain = "drain"
+	DrainerStateDrain = "drain"
+	// drainFailureRequeueDuration is the delay until we retry draining the node, such that other nodes
+	// get a chance in the queue
+	drainFailureRequeueDuration = 1 * time.Minute
+)
+
+type Reconciler struct {
+	client                 client.Client
+	guestClusterClient     client.Client
+	guestClusterKubeClient kubeclient.Interface
+	releaseProvider        releaseinfo.Provider
+	upsert.CreateOrUpdateProvider
+}
+
+func (r *Reconciler) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	log := ctrl.LoggerFrom(ctx)
+	log.Info("Reconciling")
+
+	node := &corev1.Node{}
+	err := r.guestClusterClient.Get(ctx, req.NamespacedName, node)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Info("not found", "request", req.String())
+			return ctrl.Result{}, nil
+		}
+		log.Error(err, "error getting node")
+		return ctrl.Result{}, err
+	}
+
+	if node.Annotations[DesiredDrainerAnnotationKey] != node.Annotations[LastAppliedDrainerAnnotationKey] {
+		if err = r.handleNodeDrainRequest(ctx, node, node.Annotations[DesiredDrainerAnnotationKey]); err != nil {
+			// TODO (jerzhang): properly handle failure and error detection
+			// Right now for individual failures we will endlessly retry
+			// Instead we should have some timeout (based on NodeDrainTimeout?)
+			// and error reporting to the nodepool status
+			log.Error(err, "failed to drain node")
+			return ctrl.Result{RequeueAfter: drainFailureRequeueDuration}, err
+		}
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *Reconciler) handleNodeDrainRequest(ctx context.Context, node *corev1.Node, desiredState string) error {
+	log := ctrl.LoggerFrom(ctx)
+
+	// TODO (jerzhang): this name is deterministic, but should make this a const somewhere
+	daemonPodOnNodeName := fmt.Sprintf("machine-config-daemon-%s", node.Name)
+	drainer := &drain.Helper{
+		Client:                          r.guestClusterKubeClient,
+		Force:                           true,
+		IgnoreAllDaemonSets:             true,
+		DeleteEmptyDirData:              true,
+		GracePeriodSeconds:              -1,
+		SkipWaitForDeleteTimeoutSeconds: 300,
```

Comment:
We would only want to set SkipWaitForDeleteTimeoutSeconds when the Node is unreachable to equal our other draining https://github.com/kubernetes-sigs/cluster-api/blob/main/internal/controllers/machine/machine_controller.go#L563-L566

We do only for the unreachable scenario because we acknowledge the state and accept the tradeoffs there.

https://kubernetes.io/docs/concepts/architecture/nodes/
>In some cases when the node is unreachable, the API server is unable to communicate with the kubelet on the node. The decision to delete the pods cannot be communicated to the kubelet until communication with the API server is re-established. In the meantime, the pods that are scheduled for deletion may continue to run on the partitioned node.

https://github.com/kubernetes-sigs/cluster-api/issues/1870
https://github.com/kubernetes-sigs/cluster-api/pull/2165
https://github.com/kubernetes-sigs/cluster-api/issues/1870

Reply:
ack, will change